# Titanic — uma etapa por vez, com resolução para a gravação

Este notebook reorganiza as **sete atividades do material enviado** em passos pequenos. Cada passo traz objetivo, explicação do comando, código e interpretação. As soluções estão junto das atividades para a professora mostrar na gravação.

**Como conduzir:** leia o objetivo, peça uma previsão, execute apenas a célula seguinte e comente a saída. Oriente os alunos a pausar o vídeo e repetir. Execute de cima para baixo: algumas células dependem de variáveis criadas antes. Após reiniciar o kernel, execute novamente desde o início.

**Preparação:** mantenha este notebook e o `train.csv` fornecido na mesma pasta. Abra essa pasta no VS Code e selecione um kernel com pandas. O arquivo é local; não precisamos entrar no Kaggle. Se faltar pandas, execute `%pip install pandas` em uma célula do notebook e reinicie o kernel. A instalação não faz parte da análise.

**Sugestão para a gravação:** divida em dois blocos, com uma pausa depois da atividade 3. Para iniciantes, explicar todas as sete atividades pode exigir 90–120 minutos. Se tiver apenas uma hora, priorize leitura, inspeção, filtros e agregação, deixando imputação e relatório para outro bloco.

## Antes do código: o que esta tabela representa?
Cada linha representa um passageiro **do arquivo fornecido**, não todas as pessoas a bordo. Cada coluna descreve uma característica. Dataset significa conjunto de dados. CSV é um arquivo de texto que organiza registros e campos; pandas interpreta esse arquivo como tabela.

| Coluna | Significado nesta atividade |
|---|---|
| PassengerId | Identificador do passageiro |
| Survived | 0: não sobreviveu; 1: sobreviveu |
| Pclass | Classe da passagem: 1, 2 ou 3 |
| Name | Nome registrado |
| Sex | Sexo registrado, com valores female e male |
| Age | Idade em anos; pode estar ausente |
| SibSp | Quantidade de irmãos/irmãs e cônjuges a bordo |
| Parch | Quantidade de pais/mães e filhos/filhas a bordo |
| Ticket | Código do bilhete |
| Fare | Tarifa registrada |
| Cabin | Cabine registrada; pode estar ausente |
| Embarked | Porto de embarque: C, Q ou S |

Neste CSV há **891 registros e 342 sobreviventes**. As conclusões se referem a esse conjunto.

## Vocabulário para ler o código

| Escrita | O que significa |
|---|---|
| `titanic` | Variável que guarda a tabela |
| `=` | Atribui um resultado a um nome |
| `==` | Compara dois valores |
| `pd` | Apelido usado para a biblioteca pandas |
| `titanic["Age"]` | Seleciona uma coluna; o resultado é uma Series |
| `titanic[["Age", "Fare"]]` | Seleciona várias colunas; o resultado é um DataFrame |
| `.head()` | Chama um método para mostrar o começo da tabela |
| `.shape` | Consulta um atributo; não leva parênteses |
| `(5)` em `.head(5)` | Argumento: informação entregue ao método |
| `# comentário` | Texto explicativo que Python não executa |

**Função** é uma operação que chamamos, como `len(titanic)`. **Método** é uma operação acessada por um objeto, como `titanic.head()`. O ponto acessa atributos e métodos; os parênteses chamam a operação. Os colchetes selecionam elementos ou colunas. Não são símbolos intercambiáveis.

## Preparação — trazer o arquivo para o Python


In [ ]:
Coluna O que significa	Exemplo ou interpretação
PassengerId	Identificador do passageiro. É a chave usada para ligar passageiros, viagens e resultados.	1 identifica um registro; não representa idade ou posição social.
Name	Nome registrado do passageiro.	Pode incluir sobrenome, título e informações entre parênteses.
Sex	Sexo registrado no conjunto de dados.	female = feminino; male = masculino.
Age	Idade em anos.	22 = 22 anos. Valores fracionários podem representar bebês. Ausente significa idade desconhecida.
SibSp	Quantidade de irmãos, irmãs e cônjuges a bordo.	1 indica uma pessoa dessas relações, sem informar qual delas.
Parch	Quantidade de pais, mães, filhos e filhas a bordo.	0 significa nenhum familiar dessas categorias registrado.
Pclass	Classe da passagem.	1 = primeira; 2 = segunda; 3 = terceira.
Ticket	Código do bilhete.	Pode conter letras e números. Pode aparecer para vários passageiros.
Fare	Tarifa registrada no dataset, em libras esterlinas.	Não devemos assumir que sempre representa o preço individual pago por uma pessoa.
Embarked	Código do porto de embarque.	C = Cherbourg; Q = Queenstown; S = Southampton.
Survived	Indica se o passageiro sobreviveu.	0 = não sobreviveu; 1 = sobreviveu. Ausente não significa 0.
codigo_porto	Código que identifica o porto no cadastro didático.	Corresponde ao código de Embarked, permitindo o merge.
nome_porto	Nome do porto no cadastro.	Cherbourg, Queenstown ou Southampton.
Cabin	Cabine registrada. Aparece no train.csv, mas não nas quatro tabelas separadas desta atividade.	Pode conter código de cabine ou estar ausente.

### P1. Importar pandas

**O que queremos:** Disponibilizar a biblioteca de tabelas.

**Explique antes de executar:**

`import pandas` carrega a biblioteca. `as pd` define um apelido curto; não cria uma tabela.

In [1]:
import pandas as pd

**O que observar e dizer na gravação:** Esta célula normalmente não mostra saída. Ausência de saída não significa falha.

### P2. Ler o CSV

**O que queremos:** Criar a tabela a partir de train.csv.

**Explique antes de executar:**

`pd.read_csv` é a função de leitura. `"train.csv"` é o nome do arquivo, passado como argumento. `=` guarda a tabela no nome `titanic`. O caminho é relativo à pasta de execução do notebook.

In [2]:
titanic = pd.read_csv("train.csv")

**O que observar e dizer na gravação:** Se aparecer FileNotFoundError, confira o nome do arquivo e a pasta de execução; não mude o código de análise para resolver um arquivo ausente.

### P3. Ver as primeiras linhas

**O que queremos:** Conferir se os dados foram lidos.

**Explique antes de executar:**

`head(5)` mostra as primeiras cinco linhas. Na última linha de uma célula, Jupyter exibe o resultado da expressão. O índice à esquerda começa em 0 e não é a coluna PassengerId.

In [3]:
titanic.head(5)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Thayer)",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


**O que observar e dizer na gravação:** Devem aparecer as colunas do dicionário acima. O primeiro registro tem PassengerId 1. Não estamos vendo somente cinco passageiros na base: estamos exibindo cinco.

### P4. Conferir o identificador

**O que queremos:** Checar se cada PassengerId aparece uma vez.

**Explique antes de executar:**

`is_unique` é um atributo da coluna, sem parênteses. `True` indica que os identificadores não se repetem. Essa checagem não prova que todas as demais informações estão corretas.

In [4]:
titanic["PassengerId"].is_unique

True

**O que observar e dizer na gravação:** Neste arquivo, o resultado é True.

## Atividade 1 — conhecer os dados

Mostre dimensões, nomes das colunas, tipos e quantidade e percentual de valores ausentes em cada coluna. Identifique as três colunas com mais nulos.



**Plano:** verificar tamanho → colunas → tipos → ausências → ordenar as maiores ausências.

### 1.1. Dimensões

**O que queremos:** Saber quantas linhas e colunas temos.

**Explique antes de executar:**

`.shape` retorna uma tupla: primeiro linhas, depois colunas. É atributo, não método.

In [5]:
titanic.shape

(891, 12)

**O que observar e dizer na gravação:** Resultado: (891, 12). São 891 registros e 12 campos.

### 1.2. Nomes das colunas

**O que queremos:** Conhecer os nomes exatos que podemos selecionar.

**Explique antes de executar:**

`.columns` reúne os nomes; `.tolist()` converte essa coleção em lista. Maiúsculas e minúsculas importam: Age e age não são a mesma chave.

In [6]:
titanic.columns.tolist()

['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked']

**O que observar e dizer na gravação:** Leia os nomes e relacione-os ao dicionário. Selecionar um nome inexistente causa KeyError.

### 1.3. Tipos

**O que queremos:** Entender como cada coluna foi interpretada.

**Explique antes de executar:**

`.dtypes` mostra o tipo por coluna. `int64` representa inteiros; `float64`, números com parte decimal ou ausências neste caso; `object`, aqui, aparece nas colunas de texto. Tipo numérico não garante significado quantitativo: PassengerId é identificador.

In [7]:
titanic.dtypes

,0
PassengerId,int64
Survived,int64
Pclass,int64
Name,object
Sex,object
Age,float64
SibSp,int64
Parch,int64
Ticket,object
Fare,float64


**O que observar e dizer na gravação:** Age e Fare são float64. Survived é numérica, mas representa duas categorias.

### 1.4. Detectar ausências

**O que queremos:** Marcar cada célula sem valor.

**Explique antes de executar:**

`.isna()` devolve uma tabela booleana: True se o valor está ausente, False se existe. Ausente não é a mesma coisa que zero ou string vazia.

In [8]:
titanic.isna().head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,False,False,False,False,False,False,False,False,False,False,True,False
1,False,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,True,False
3,False,False,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,True,False


**O que observar e dizer na gravação:** A ausência de Cabin aparece como True em várias linhas. Esta saída marca posições; ainda não conta.

### 1.5. Contar ausências

**O que queremos:** Somar as marcações por coluna.

**Explique antes de executar:**

`.sum()` na tabela booleana conta True como 1 e False como 0. O padrão soma por coluna. `.to_frame("quantidade")` transforma a Series em uma tabela de uma coluna com esse nome.

In [9]:
nulos = titanic.isna().sum().to_frame("quantidade")
nulos

,quantidade
PassengerId,0
Survived,0
Pclass,0
Name,0
Sex,0
Age,177
SibSp,0
Parch,0
Ticket,0
Fare,0


**O que observar e dizer na gravação:** Age tem 177 ausências; Cabin, 687; Embarked, 2.

### 1.6. Calcular percentuais

**O que queremos:** Relacionar as ausências ao total de linhas.

**Explique antes de executar:**

`len(titanic)` devolve 891. Dividimos quantidade por total e multiplicamos por 100. A atribuição a `nulos["percentual"]` cria uma nova coluna no resumo.

In [10]:
nulos["percentual"] = nulos["quantidade"] / len(titanic) * 100
nulos

,quantidade,percentual
PassengerId,0,0.000000
Survived,0,0.000000
Pclass,0,0.000000
Name,0,0.000000
Sex,0,0.000000
Age,177,19.865320
SibSp,0,0.000000
Parch,0,0.000000
Ticket,0,0.000000
Fare,0,0.000000


**O que observar e dizer na gravação:** Age tem aproximadamente 19,87% de ausências; Cabin, 77,10%; Embarked, 0,22%. O denominador é 891 em todas as colunas.

### 1.7. Ordenar e destacar

**O que queremos:** Encontrar as três colunas mais incompletas.

**Explique antes de executar:**

`sort_values` ordena pelo campo indicado. `ascending=False` pede ordem decrescente. Guardamos a versão ordenada e usamos head(3).

In [11]:
nulos_ordenados = nulos.sort_values("quantidade", ascending=False)
nulos_ordenados.head(3)

,quantidade,percentual
Cabin,687,77.104377
Age,177,19.865320
Embarked,2,0.224467


**O que observar e dizer na gravação:** Cabin, Age e Embarked. São informações que exigem atenção antes de análises; não é uma ordem de importância das colunas.

## Atividade 2 — filtros combinados

**Enunciado:** mulheres da primeira ou segunda classe, idade de pelo menos 18 e tarifa acima de 30. Mostrar ID, nome, idade, classe e tarifa; contar o resultado.

**Plano:** escrever uma condição por vez → combinar → escolher as colunas → interpretar exclusões.

### 2.1. Comparar uma coluna

**O que queremos:** Marcar os registros com Sex igual a female.

**Explique antes de executar:**

`==` compara cada valor da coluna com a string. A saída tem um True ou False por linha. Isso é uma máscara booleana, não a tabela filtrada.

In [12]:
sexo_feminino = titanic["Sex"] == "female"
sexo_feminino.head()

,Sex
0,False
1,True
2,True
3,True
4,False


**O que observar e dizer na gravação:** No início, aparecem False, True, True, True, False. A variável guarda a condição para reutilizar.

### 2.2. Aceitar duas classes

**O que queremos:** Marcar classe 1 ou classe 2.

**Explique antes de executar:**

`.isin([1, 2])` verifica se cada valor pertence à lista. Os colchetes internos criam a lista de opções.

In [13]:
classe_1_ou_2 = titanic["Pclass"].isin([1, 2])
classe_1_ou_2.head()

,Pclass
0,False
1,True
2,False
3,True
4,False


**O que observar e dizer na gravação:** Classe 3 recebe False. isin permite várias opções na mesma condição.

### 2.3. Idade e tarifa

**O que queremos:** Registrar as duas condições numéricas.

**Explique antes de executar:**

`>= 18` inclui quem tem exatamente 18; `> 30` exclui quem tem exatamente 30. Comparações com NaN de Age retornam False nesta coluna. Não estamos afirmando a idade dos registros ausentes.

In [14]:
idade_minima = titanic["Age"] >= 18
tarifa_acima_30 = titanic["Fare"] > 30
idade_minima.head()

,Age
0,True
1,True
2,True
3,True
4,True


**O que observar e dizer na gravação:** Uma idade desconhecida não atende à condição, porque não podemos confirmar que é pelo menos 18.

### 2.4. Combinar as condições

**O que queremos:** Exigir que todas sejam verdadeiras na mesma linha.

**Explique antes de executar:**

`&` é o E entre máscaras pandas. Não use o `and` do Python entre Series. Em expressões diretas, coloque cada comparação entre parênteses. Aqui as variáveis já guardam máscaras prontas.

In [15]:
filtro = sexo_feminino & classe_1_ou_2 & idade_minima & tarifa_acima_30
filtro.sum()

np.int64(75)

**O que observar e dizer na gravação:** 75 registros atendem a todas as condições.

### 2.5. Selecionar linhas e colunas

**O que queremos:** Construir a tabela que o enunciado pede.

**Explique antes de executar:**

`.loc[linhas, colunas]` recebe a máscara antes da vírgula e a lista de colunas depois. A seleção preserva os índices originais.

In [16]:
selecionadas = titanic.loc[filtro, ["PassengerId", "Name", "Age", "Pclass", "Fare"]]
selecionadas

,PassengerId,Name,Age,Pclass,Fare
1,2,"Cumings, Mrs. John Bradley (Florence Briggs Thayer)",38.0,1,71.2833
3,4,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",35.0,1,53.1000
52,53,"Harper, Mrs. Henry Sleeper (Myna Haxtun)",49.0,1,76.7292
61,62,"Icard, Miss. Amelie",38.0,1,80.0000
88,89,"Fortune, Miss. Mabel Helen",23.0,1,263.0000
151,152,"Pears, Mrs. Thomas (Edith Wearne)",22.0,1,66.6000
195,196,"Lurette, Miss. Elise",58.0,1,146.5208
215,216,"Newell, Miss. Madeleine",31.0,1,113.2750
218,219,"Bazzani, Miss. Albina",32.0,1,76.2917
230,231,"Harris, Mrs. Henry Birkhardt (Irene Wallach)",35.0,1,83.4750


**O que observar e dizer na gravação:** Mostre a tabela. Se a saída for abreviada, isso é uma configuração de exibição, não perda de registros.

### 2.6. Contar a seleção

**O que queremos:** Responder a quantidade explicitamente.

**Explique antes de executar:**

`len` conta linhas do resultado. `print` mostra uma mensagem junto do número.

In [17]:
print("Selecionadas:", len(selecionadas))

Selecionadas: 75


**O que observar e dizer na gravação:** Resultado: 75. Não preenchemos idades ausentes para forçar inclusão no filtro.

## Atividade 3 — sobrevivência por sexo e classe

**Enunciado:** construir passageiros, sobreviventes e taxa percentual por sexo e classe, ordenar por taxa e conferir o total.

**Ideia:** dividir em grupos → calcular dentro de cada grupo → comparar os resumos.

### 3.1. Entender a média de 0 e 1

**O que queremos:** Ver por que mean pode medir proporção.

**Explique antes de executar:**

Nos valores [0, 1, 1, 0], há dois sobreviventes em quatro pessoas. Soma = 2; média = 2/4 = 0,5. Multiplicar por 100 dá 50%. Isso funciona porque o indicador usa 0 e 1.

In [18]:
exemplo = pd.Series([0, 1, 1, 0])
print("Soma:", exemplo.sum())
print("Média:", exemplo.mean())
print("Percentual:", exemplo.mean() * 100)

Soma: 2
Média: 0.5
Percentual: 50.0


**O que observar e dizer na gravação:** Uma média de 0,5 aqui não é meio passageiro: é uma proporção de sobrevivência.

### 3.2. Agrupar e resumir

**O que queremos:** Produzir uma linha por combinação de sexo e classe.

**Explique antes de executar:**

`groupby(["Sex", "Pclass"])` define os grupos. `.agg(...)` calcula estatísticas. Na escrita `passageiros=("PassengerId", "size")`, passageiros é o nome de saída; PassengerId é a coluna de entrada; size conta as linhas. sum soma o indicador; mean calcula a proporção.

In [19]:
resumo = titanic.groupby(["Sex", "Pclass"]).agg(
    passageiros=("PassengerId", "size"),
    sobreviventes=("Survived", "sum"),
    taxa=("Survived", "mean")
)
resumo

passageiros  sobreviventes      taxa
Sex    Pclass                                      
female 1                94             91  0.968085
       2                76             70  0.921053
       3               144             72  0.500000
male   1               122             45  0.368852
       2               108             17  0.157407
       3               347             47  0.135447

**O que observar e dizer na gravação:** Temos seis grupos. Os nomes dos grupos aparecem no índice; taxa ainda está entre 0 e 1.

### 3.3. Transformar o índice em colunas

**O que queremos:** Deixar o resumo com colunas comuns.

**Explique antes de executar:**

`.reset_index()` transforma os identificadores dos grupos em colunas e cria um índice simples. Não altera a quantidade de passageiros.

In [20]:
resumo = resumo.reset_index()
resumo

,Sex,Pclass,passageiros,sobreviventes,taxa
0,female,1,94,91,0.968085
1,female,2,76,70,0.921053
2,female,3,144,72,0.500000
3,male,1,122,45,0.368852
4,male,2,108,17,0.157407
5,male,3,347,47,0.135447


**O que observar e dizer na gravação:** Sex e Pclass agora podem ser acessadas como outras colunas.

### 3.4. Percentual e ordenação

**O que queremos:** Comparar grupos pela taxa.

**Explique antes de executar:**

Crie taxa_percentual multiplicando a proporção por 100. Ordene pela nova coluna em ordem decrescente.

In [21]:
resumo["taxa_percentual"] = resumo["taxa"] * 100
resumo.sort_values("taxa_percentual", ascending=False)

,Sex,Pclass,passageiros,sobreviventes,taxa,taxa_percentual
0,female,1,94,91,0.968085,96.808511
1,female,2,76,70,0.921053,92.105263
2,female,3,144,72,0.500000,50.000000
3,male,1,122,45,0.368852,36.885246
4,male,2,108,17,0.157407,15.740741
5,male,3,347,47,0.135447,13.544669


**O que observar e dizer na gravação:** Neste arquivo, mulheres da classe 1 têm aproximadamente 96,81%; homens da classe 3, 13,54%. São associações observadas, sem demonstração de causa.

### 3.5. Conferir o total

**O que queremos:** Verificar que as contagens somam o tamanho da base.

**Explique antes de executar:**

`.sum()` agora soma uma coluna numérica do resumo. Comparamos ao total original usando ==.

In [22]:
print("Total nos grupos:", resumo["passageiros"].sum())
print("Total na base:", len(titanic))
print("Conferência:", resumo["passageiros"].sum() == len(titanic))

Total nos grupos: 891
Total na base: 891
Conferência: True


**O que observar e dizer na gravação:** 891 em ambos; True na conferência. groupby pode excluir chaves ausentes por padrão; aqui Sex e Pclass não têm ausências.

## Pausa sugerida na gravação

Recapitule: lemos a base, conhecemos os dados, filtramos linhas e resumimos grupos. Peça aos alunos que reproduzam as três atividades antes de continuar.

## Atividade 4 — investigar e preencher idades ausentes

**Enunciado:** comparar registros com idade conhecida e ausente; preencher numa cópia usando mediana por sexo e classe, manter Age original e marcar imputações.

**Imputar** significa preencher uma ausência com um valor estimado por uma regra. Não significa recuperar a idade verdadeira. **Mediana** é o valor central dos valores ordenados; com quantidade par, é a média dos dois centrais.

### 4.1. Criar cópia e indicador

**O que queremos:** Separar o experimento de preenchimento da tabela original.

**Explique antes de executar:**

`.copy()` cria outra tabela. `.isna()` na coluna Age devolve a marcação de ausência. Guardamos essa marcação antes de preencher.

In [23]:
idades = titanic.copy()
idades["idade_ausente"] = idades["Age"].isna()
idades[["Age", "idade_ausente"]].head(10)

,Age,idade_ausente
0,22.0,False
1,38.0,False
2,26.0,False
3,35.0,False
4,35.0,False
5,NaN,True
6,54.0,False
7,2.0,False
8,27.0,False
9,14.0,False


**O que observar e dizer na gravação:** True significa que a idade original não estava disponível; False significa que havia idade registrada.

### 4.2. Comparar os dois grupos

**O que queremos:** Ver contagem e taxa segundo presença de idade.

**Explique antes de executar:**

Agrupe por idade_ausente. False reúne idades conhecidas; True reúne desconhecidas. A taxa considera Survived, que está preenchida neste arquivo.

In [24]:
comparacao_idades = idades.groupby("idade_ausente").agg(
    passageiros=("PassengerId", "size"),
    taxa=("Survived", "mean")
)
comparacao_idades["taxa_percentual"] = comparacao_idades["taxa"] * 100
comparacao_idades

,passageiros,taxa,taxa_percentual
idade_ausente,,,
False,714,0.406162,40.616246
True,177,0.293785,29.378531


**O que observar e dizer na gravação:** 714 idades conhecidas, com taxa de aproximadamente 40,62%; 177 ausentes, com 29,38%. Isso descreve grupos, não prova que ausência de idade afeta sobrevivência.

### 4.3. Calcular medianas dos grupos

**O que queremos:** Conhecer o valor que será usado para cada grupo.

**Explique antes de executar:**

Agrupe por Sex e Pclass, selecione Age e calcule median(). A mediana ignora idades ausentes.

In [25]:
idades.groupby(["Sex", "Pclass"])["Age"].median()

Sex     Pclass
female  1         35.0
        2         28.0
        3         21.5
male    1         40.0
        2         30.0
        3         25.0
Name: Age, dtype: float64

**O que observar e dizer na gravação:** As seis medianas são: mulheres classes 1/2/3 = 35, 28, 21,5; homens classes 1/2/3 = 40, 30, 25.

### 4.4. Alinhar a mediana a cada passageiro

**O que queremos:** Obter uma mediana de grupo para cada linha da base.

**Explique antes de executar:**

`.transform("median")` calcula a mediana dentro do grupo e repete o resultado nas linhas daquele grupo. Diferente da agregação, mantém 891 posições alinhadas aos passageiros.

In [26]:
mediana_grupo = idades.groupby(["Sex", "Pclass"])["Age"].transform("median")
mediana_grupo.head(10)

,Age
0,25.0
1,35.0
2,21.5
3,35.0
4,25.0
5,25.0
6,40.0
7,25.0
8,21.5
9,28.0


**O que observar e dizer na gravação:** Não temos seis linhas, mas uma sugestão por passageiro. Esse alinhamento permite preencher sem merge.

### 4.5. Preencher numa nova coluna

**O que queremos:** Substituir apenas as ausências, preservando Age.

**Explique antes de executar:**

`.fillna(mediana_grupo)` mantém valores existentes e usa a mediana correspondente somente onde havia ausência. Guardamos em Age_imputada.

In [27]:
idades["Age_imputada"] = idades["Age"].fillna(mediana_grupo)
idades.loc[idades["idade_ausente"], ["PassengerId", "Age", "Age_imputada"]].head()

,PassengerId,Age,Age_imputada
5,6,NaN,25.0
17,18,NaN,30.0
19,20,NaN,21.5
26,27,NaN,25.0
28,29,NaN,21.5


**O que observar e dizer na gravação:** Age continua ausente nas linhas mostradas; Age_imputada tem estimativas. idade_ausente também registra quais valores foram imputados neste arquivo.

### 4.6. Conferir o preenchimento

**O que queremos:** Ver nulos restantes e preservação do original.

**Explique antes de executar:**

`.equals(...)` compara a Series original à da cópia. Se um grupo não tivesse idades conhecidas, sua mediana seria ausente e o preenchimento não resolveria tudo.

In [28]:
print("Nulos após preencher:", idades["Age_imputada"].isna().sum())
print("Age original preservada:", idades["Age"].equals(titanic["Age"]))

Nulos após preencher: 0
Age original preservada: True


**O que observar e dizer na gravação:** Neste CSV: 0 nulos restantes e True para preservação. Em modelagem, estatísticas de imputação devem ser aprendidas somente nos dados de treino de cada divisão; aqui a atividade é descritiva.

## Atividade 5 — familiares a bordo

**Enunciado:** criar tamanho_familia = SibSp + Parch + 1 e viajava_sozinho; comparar passageiros, sobreviventes e taxas.

A regra usa familiares registrados nos campos. Não identifica amigos nem todos os acompanhantes; bilhete compartilhado não entra nesta definição.

### 5.1. Calcular tamanho de família

**O que queremos:** Somar familiares e a própria pessoa.

**Explique antes de executar:**

As somas entre colunas são feitas linha a linha. `+ 1` inclui o próprio passageiro.

In [29]:
familias = titanic.copy()
familias["tamanho_familia"] = familias["SibSp"] + familias["Parch"] + 1
familias[["SibSp", "Parch", "tamanho_familia"]].head()

,SibSp,Parch,tamanho_familia
0,1,0,2
1,1,0,2
2,0,0,1
3,1,0,2
4,0,0,1


**O que observar e dizer na gravação:** SibSp 0 e Parch 0 produzem tamanho 1. Não somamos todos os passageiros em uma única família.

### 5.2. Marcar ausência de familiares

**O que queremos:** Identificar tamanho_familia igual a 1.

**Explique antes de executar:**

`.eq(1)` faz a mesma comparação que `== 1`. O nome viajava_sozinho é operacional: significa sem familiares registrados por essa regra.

In [30]:
familias["viajava_sozinho"] = familias["tamanho_familia"].eq(1)
familias[["tamanho_familia", "viajava_sozinho"]].head()

,tamanho_familia,viajava_sozinho
0,2,False
1,2,False
2,1,True
3,2,False
4,1,True


**O que observar e dizer na gravação:** True marca ausência de familiares registrados; não comprova que não havia companhia.

### 5.3. Comparar os grupos

**O que queremos:** Resumir sobrevivência segundo essa definição.

**Explique antes de executar:**

Reutilize groupby e agg. Adicionamos percentual para facilitar a leitura; a proporção continua em taxa.

In [31]:
resumo_familias = familias.groupby("viajava_sozinho").agg(
    passageiros=("PassengerId", "size"),
    sobreviventes=("Survived", "sum"),
    taxa=("Survived", "mean")
)
resumo_familias["taxa_percentual"] = resumo_familias["taxa"] * 100
resumo_familias

,passageiros,sobreviventes,taxa,taxa_percentual
viajava_sozinho,,,,
False,354,179,0.505650,50.564972
True,537,163,0.303538,30.353818


**O que observar e dizer na gravação:** Sem familiares: 537 passageiros, 163 sobreviventes, aproximadamente 30,35%. Com familiares: 354, 179, aproximadamente 50,56%. Não conclua que companhia causou sobrevivência.

## Atividade 6 — faixas etárias

**Enunciado:** usar Age original para formar menos de 12, 12 a menos de 18, 18 a menos de 60 e 60 ou mais; manter ausentes como Não informado.

**Limites:** 12 entra na segunda faixa; 18, na terceira; 60, na quarta. Este arquivo não tem idades negativas.

### 6.1. Criar intervalos

**O que queremos:** Converter idades numéricas em categorias.

**Explique antes de executar:**

`pd.cut` recebe a coluna, limites em bins e nomes em labels. `right=False` inclui o limite esquerdo e exclui o direito: [12,18), por exemplo. `float("inf")` é infinito positivo, para não impor teto à última faixa.

In [32]:
faixas = titanic.copy()
faixas["faixa_etaria"] = pd.cut(
    faixas["Age"],
    bins=[0, 12, 18, 60, float("inf")],
    right=False,
    labels=["Menos de 12", "12 a menos de 18", "18 a menos de 60", "60 ou mais"]
)
faixas[["Age", "faixa_etaria"]].head(10)

,Age,faixa_etaria
0,22.0,18 a menos de 60
1,38.0,18 a menos de 60
2,26.0,18 a menos de 60
3,35.0,18 a menos de 60
4,35.0,18 a menos de 60
5,NaN,NaN
6,54.0,18 a menos de 60
7,2.0,Menos de 12
8,27.0,18 a menos de 60
9,14.0,12 a menos de 18


**O que observar e dizer na gravação:** Idades ausentes continuam sem categoria. Não usamos Age_imputada, para não misturar estimativa com observação.

### 6.2. Registrar a categoria de ausência

**O que queremos:** Dar um nome visível às idades desconhecidas.

**Explique antes de executar:**

cut produz uma coluna categórica. Antes de preencher com um novo rótulo, adicionamos esse rótulo às categorias permitidas via `.cat.add_categories`. Depois fillna usa o rótulo.

In [33]:
faixas["faixa_etaria"] = faixas["faixa_etaria"].cat.add_categories("Não informado")
faixas["faixa_etaria"] = faixas["faixa_etaria"].fillna("Não informado")
faixas.loc[faixas["Age"].isna(), ["Age", "faixa_etaria"]].head()

,Age,faixa_etaria
5,NaN,Não informado
17,NaN,Não informado
19,NaN,Não informado
26,NaN,Não informado
28,NaN,Não informado


**O que observar e dizer na gravação:** A idade original segue ausente. Somente a coluna de faixa ganha o texto Não informado.

### 6.3. Comparar faixas

**O que queremos:** Contar e calcular taxa por categoria.

**Explique antes de executar:**

`observed=True` pede somente categorias que aparecem nos dados. Com a categoria Não informado, os 177 passageiros sem idade são incluídos no resumo.

In [34]:
resumo_faixas = faixas.groupby("faixa_etaria", observed=True).agg(
    passageiros=("PassengerId", "size"),
    taxa=("Survived", "mean")
)
resumo_faixas["taxa_percentual"] = resumo_faixas["taxa"] * 100
resumo_faixas

,passageiros,taxa,taxa_percentual
faixa_etaria,,,
Menos de 12,68,0.573529,57.352941
12 a menos de 18,45,0.488889,48.888889
18 a menos de 60,575,0.386087,38.608696
60 ou mais,26,0.269231,26.923077
Não informado,177,0.293785,29.378531


**O que observar e dizer na gravação:** Contagens: 68, 45, 575, 26 e 177, respectivamente. Somam 891. Compare os tamanhos antes de interpretar diferenças de taxas.

## Atividade 7 — relatório final

**Enunciado:** por classe e presença de familiares, produzir passageiros, sobreviventes, taxa percentual, mediana de tarifa e idade e número de idades conhecidas. Destacar grupos com pelo menos 20 registros e taxa acima da geral.

**Plano:** criar a chave de familiares → resumir → converter taxa → calcular referência geral → filtrar → escrever observações.

### 7.1. Preparar a chave de familiares

**O que queremos:** Criar a definição usada no relatório.

**Explique antes de executar:**

Somar SibSp e Parch e comparar a zero equivale a verificar tamanho_familia igual a 1. Usamos outra cópia para o relatório.

In [35]:
relatorio_base = titanic.copy()
relatorio_base["viajava_sozinho"] = (relatorio_base["SibSp"] + relatorio_base["Parch"]).eq(0)
relatorio_base[["Pclass", "viajava_sozinho"]].head()

,Pclass,viajava_sozinho
0,3,False
1,1,False
2,3,True
3,1,False
4,3,True


**O que observar e dizer na gravação:** Temos as duas chaves para os grupos: classe e indicação de ausência de familiares.

### 7.2. Calcular todas as métricas

**O que queremos:** Formar uma tabela com uma linha por grupo.

**Explique antes de executar:**

`size` conta todas as linhas; `count` em Age conta somente valores conhecidos. `median` em Age ignora ausências. Mediana de tarifa resume a tarifa registrada; não demonstra quanto cada pessoa efetivamente pagou em todas as situações.

In [36]:
relatorio = relatorio_base.groupby(["Pclass", "viajava_sozinho"]).agg(
    passageiros=("PassengerId", "size"),
    sobreviventes=("Survived", "sum"),
    taxa=("Survived", "mean"),
    mediana_tarifa=("Fare", "median"),
    mediana_idade=("Age", "median"),
    idades_conhecidas=("Age", "count")
).reset_index()
relatorio["taxa_percentual"] = relatorio["taxa"] * 100
relatorio

,Pclass,viajava_sozinho,passageiros,sobreviventes,taxa,mediana_tarifa,mediana_idade,idades_conhecidas,taxa_percentual
0,1,False,107,78,0.728972,79.6500,36.0,100,72.897196
1,1,True,109,58,0.532110,31.0000,38.5,86,53.211009
2,2,False,80,51,0.637500,26.0000,27.0,80,63.750000
3,2,True,104,36,0.346154,13.0000,31.0,93,34.615385
4,3,False,167,50,0.299401,18.0000,18.0,130,29.940120
5,3,True,324,69,0.212963,7.8958,26.0,225,21.296296


**O que observar e dizer na gravação:** Se passageiros e idades_conhecidas diferem, a mediana de idade foi calculada sobre um subconjunto. Existem seis grupos neste arquivo.

### 7.3. Calcular taxa geral

**O que queremos:** Criar a referência para comparação.

**Explique antes de executar:**

`Survived.mean()` calcula sobreviventes dividido pelo total neste CSV. Guardamos como proporção e mostramos em percentual.

In [37]:
taxa_geral = titanic["Survived"].mean()
print("Taxa geral (%):", round(taxa_geral * 100, 2))

Taxa geral (%): 38.38


**O que observar e dizer na gravação:** 342 / 891: aproximadamente 38,38%. round(..., 2) arredonda a exibição para duas casas.

### 7.4. Destacar grupos

**O que queremos:** Escolher grupos suficientemente grandes e acima da taxa geral.

**Explique antes de executar:**

`>= 20` inclui grupos com exatamente 20. `> taxa_geral` exige taxa estritamente maior. Use proporção com proporção, sem comparar taxa_percentual com taxa_geral.

In [38]:
destaque = (relatorio["passageiros"] >= 20) & (relatorio["taxa"] > taxa_geral)
grupos_destacados = relatorio.loc[destaque]
grupos_destacados

,Pclass,viajava_sozinho,passageiros,sobreviventes,taxa,mediana_tarifa,mediana_idade,idades_conhecidas,taxa_percentual
0,1,False,107,78,0.728972,79.65,36.0,100,72.897196
1,1,True,109,58,0.532110,31.00,38.5,86,53.211009
2,2,False,80,51,0.637500,26.00,27.0,80,63.750000


**O que observar e dizer na gravação:** São três grupos: classe 1 com familiares, classe 1 sem familiares e classe 2 com familiares. O corte 20 vem do exercício, não de uma garantia estatística.

### 7.5. Conferir total do relatório

**O que queremos:** Garantir coerência das contagens.

**Explique antes de executar:**

A soma das contagens dos grupos deve recuperar a base. Essa checagem é uma evidência de consistência, não uma prova de qualidade completa.

In [39]:
print("Passageiros no relatório:", relatorio["passageiros"].sum())
print("Sobreviventes no relatório:", relatorio["sobreviventes"].sum())

Passageiros no relatório: 891
Sobreviventes no relatório: 342


**O que observar e dizer na gravação:** 891 passageiros e 342 sobreviventes.

### 7.6. Resposta escrita — observações e limitação

**Observação 1:** no arquivo fornecido, a classe 1 apresenta taxa de sobrevivência acima da taxa geral tanto entre pessoas com familiares registrados quanto entre pessoas sem familiares registrados. O grupo com familiares tem taxa aproximada de 72,90%; o grupo sem familiares, 53,21%.

**Observação 2:** na classe 2, o grupo com familiares registrados apresenta aproximadamente 63,75%, contra 34,62% no grupo sem familiares registrados. O resumo descreve uma diferença, mas ainda não separa possíveis efeitos de idade, sexo e outras características.

**Limitação:** o arquivo contém 891 registros, não todas as pessoas a bordo; 177 idades estão ausentes, e a definição de companhia usa apenas SibSp e Parch. A análise é descritiva e não demonstra que classe ou familiares causaram a sobrevivência.

**Fala sugerida:** “Um insight precisa conectar o número a uma pergunta, indicar qual grupo estamos descrevendo e respeitar o que os dados permitem concluir.”

## Fechamento — o que aprendemos?

1. `read_csv` trouxe o arquivo para uma tabela.
2. `shape`, `columns`, `dtypes` e `isna` ajudaram a conhecê-la.
3. Máscaras e `loc` selecionaram registros.
4. `groupby` e `agg` resumiram grupos.
5. `transform` devolveu uma estatística por linha para imputação.
6. `cut` criou intervalos de idade.
7. Uma conclusão combina resultado, contexto e limitação.

**Perguntas para pausar a gravação:**

- Por que uma idade ausente não é zero? **Porque desconhecimento não é uma idade observada.**
- Qual a diferença entre size e count em Age? **size conta linhas; count conta idades conhecidas.**
- Preencher com a mediana revela a idade real? **Não, é uma estimativa.**
- Por que multiplicamos mean de Survived por 100? **Para expressar a proporção como percentual.**
- Uma diferença entre grupos prova causa? **Não.**

## Exercícios adicionais para praticar — com resolução
Tente cada pergunta antes de executar sua solução. Estas atividades são complementares às sete do material original.


### A. Passageiros com idade conhecida abaixo de 18

**O que queremos:** Selecionar IDs, nomes e idades dos menores de 18 registrados.

**Explique antes de executar:**

Combine `notna()` para explicitar idade conhecida com `< 18`. A comparação por si só já exclui NaN nesta coluna; a condição adicional deixa a intenção clara.

In [40]:
menores = titanic.loc[titanic["Age"].notna() & (titanic["Age"] < 18), ["PassengerId", "Name", "Age"]]
print("Quantidade:", len(menores))
menores.head()

Quantidade: 113


,PassengerId,Name,Age
7,8,"Palsson, Master. Gosta Leonard",2.0
9,10,"Nasser, Mrs. Nicholas (Adele Achem)",14.0
10,11,"Sandstrom, Miss. Marguerite Rut",4.0
14,15,"Vestrom, Miss. Hulda Amanda Adolfina",14.0
16,17,"Rice, Master. Eugene",2.0


**O que observar e dizer na gravação:** 113 passageiros. Não afirmamos que os 177 com idade desconhecida eram adultos.

### B. Contagem por porto, incluindo ausentes

**O que queremos:** Contar todos os registros de Embarked.

**Explique antes de executar:**

`.value_counts(dropna=False)` conta cada valor e inclui a ausência; o padrão excluiria ausentes.

In [41]:
titanic["Embarked"].value_counts(dropna=False)

,count
Embarked,
S,644
C,168
Q,77
NaN,2


**O que observar e dizer na gravação:** S: 644; C: 168; Q: 77; ausentes: 2. Somam 891.

### C. Taxa por classe

**O que queremos:** Praticar groupby com uma única chave.

**Explique antes de executar:**

Conte passageiros e some sobreviventes; calcule percentual explicitamente como sobreviventes / passageiros × 100. É a mesma proporção obtida pela média do indicador 0/1.

In [42]:
por_classe = titanic.groupby("Pclass").agg(
    passageiros=("PassengerId", "size"),
    sobreviventes=("Survived", "sum")
)
por_classe["taxa_percentual"] = por_classe["sobreviventes"] / por_classe["passageiros"] * 100
por_classe

,passageiros,sobreviventes,taxa_percentual
Pclass,,,
1,216,136,62.962963
2,184,87,47.282609
3,491,119,24.236253


**O que observar e dizer na gravação:** Classes 1/2/3: 216/184/491 passageiros e taxas de aproximadamente 62,96%/47,28%/24,24%.

## Verificações finais da professora

`assert` verifica uma condição: se for verdadeira, continua; se for falsa, interrompe com AssertionError. Estas checagens conferem números e preservação do dado após todas as etapas.

In [43]:
assert titanic.shape == (891, 12)
assert titanic["PassengerId"].is_unique
assert titanic["Survived"].sum() == 342
assert titanic["Age"].isna().sum() == 177
assert len(selecionadas) == 75
assert resumo["passageiros"].sum() == len(titanic)
assert idades["Age"].equals(titanic["Age"])
assert idades["Age_imputada"].isna().sum() == 0
assert resumo_familias["passageiros"].sum() == len(titanic)
assert resumo_faixas["passageiros"].sum() == len(titanic)
assert relatorio["passageiros"].sum() == len(titanic)
assert len(grupos_destacados) == 3
assert len(menores) == 113
print("Conferências concluídas.")

Conferências concluídas.


## Fonte e escopo
Material adaptado do notebook `titanic_tabela_completa(1).ipynb` enviado pela professora e executado com o `train.csv` anexado. O notebook original informa uma cópia pública do conjunto Titanic; esta aula usa exclusivamente o arquivo local. Não treinamos modelo preditivo nem precisamos baixar dados para executá-la.